### ISSUE TITLE
[ENH] Add option for dropping epochs where there are more bads than n_interpolates_

### ISSUE THREAD

**Issue by alexrockhill** [autoreject/autoreject#266]
I can write a PR for this but I wanted to see what others think cc @jasmainak.

I think when you interpolate only a sub-selection of the channels that are bad but accept the epoch, that's a bit odd. This is because that same channel would be interpolated in another epoch if there were fewer epochs that were bad. I propose to add an option to drop these epochs.

Before
![image](https://user-images.githubusercontent.com/13473576/161659384-ca2514ca-dc5a-4a83-b3c7-04e7cd9eeb33.png)

After
![image](https://user-images.githubusercontent.com/13473576/161659306-83b6eb5a-86cf-4e4c-a26c-4ab5125dde0c.png)

Note the columns where there are some blue and some red, I think these should either all be interpolated or the epoch should be dropped (at least that should be an option).

**Comment by jasmainak** [comment]
This is on purpose. If fewer channels are bad in another epoch, that channel can be reconstructed. But if too many channels are bad, then the interpolation quality will be bad ...

**Comment by agramfort** [comment]
to me it becomes an empirical question on some data where a new option
appears useful (ideally quantitatively)

**Comment by alexrockhill** [comment]
I think whether the interpolation is good is a lot more local question; if the bad channels are adjacent, I might not want to interpolate. That could add some more nuance. I think a bit of post hoc adjustments would be nice, it could even be it's own function.

**Comment by alexrockhill** [comment]
Here's what I'm going to use on my data to be pretty picky in what I include

```
# fix:: no bad channels in good epochs
reject_log.bad_epochs = np.any(reject_log.labels == 1, axis=1)
# fix:: don't interpolate if bad channels are adjacent
adjacency, ch_names = \
    mne.channels.find_ch_adjacency(epochs.info, 'eeg')
bads = np.logical_or(reject_log.labels == 1, reject_log.labels == 2)
for i in range(len(epochs)):
    if reject_log.bad_epochs[i]:  # already bad
        continue
    bad_idxs = set(np.where(bads[i])[0])
    for bad_idx in bad_idxs:
        for bad_idx2 in bad_idxs.difference(set([bad_idx])):
            if adjacency[bad_idx, bad_idx2]:
                reject_log.bad_epochs[i] = True
```

My goal with this is to interpolate things like electrode pops which happen to one channel in one epoch or periods of increased noise in a channel over a few epochs but to exclude all other epochs where more than one or two things are going wrong. I'm happy to add a PR to this extent that can either add it to `autoreject.transform` or make an entirely separate function but I'll defer as to whether it's useful to you @jasmainak and @agramfort.

**Comment by jasmainak** [comment]
I like the idea of using adjacency to decide what channels to interpolate :) Unfortunately I just don't know "how much" better any of these alternative approaches are. So far we have avoided making ad hoc changes to the core algorithm without quantitative comparison. I'm not completely opposed to adding an option as long as the default is the original algorithm but what does the boss (@agramfort ) say?

**Comment by agramfort** [comment]
>
> I like the idea of using adjacency to decide what channels to interpolate
> :)
>

we even discussed this idea during your PhD @jasmainak ;)


> Unfortunately I just don't know "how much" better any of these alternative
> approaches are. So far we have avoided making ad hoc changes to the core
> algorithm without quantitative comparison. I'm not completely opposed to
> adding an option as long as the default is the original algorithm but what
> does the boss ***@***.*** <https://github.com/agramfort> ) say?
>
+1

would you do the improvement with some new param in init or transform? what
would be the API like?

**Comment by alexrockhill** [comment]
Maybe we could add the quantitative comparison in an example? It looks like you used HCP data in the paper but maybe we can do a smaller dataset?

I think the API could be something like `autoreject.transform(epochs, interp_adjacent=False).

**Comment by alexrockhill** [comment]
A really good example of the issue is that I get some like this

![image](https://user-images.githubusercontent.com/13473576/161830124-fb97641d-75e4-4315-b660-d668f9d46d0d.png)

Where a channel is mostly bad but about half are kept as false positives and half are interpolated; this is the main internal consistency issue to me.

**Comment by jasmainak** [comment]
can you explain in English (not Python :) what `interp_adjacent` would do? That is what is the rule for interpolating/dropping? If any of the adjacent channels are bad, you don't interpolate regardless of `rho` ? Do you drop then?

**Comment by alexrockhill** [comment]
If `interp_adjacent` is `True`, then the channels will be interpolated normally (current behavior), if `False`, then if a channel is to be interpolated but is adjacent to another channel that is bad or interpolated, the epoch would be dropped instead.

**Comment by jasmainak** [comment]
no objection to implementing this but one thought that comes to mind is that this behavior might be too aggressive and drop too many epochs. You can probably still interpolate fine even if one neighboring channel is bad ...

**Comment by alexrockhill** [comment]
Agreed, the default could be `False` but if you want to be careful with interpolation, you could set it to `True`.

**Comment by jasmainak** [comment]
sounds good, PR welcome !

**Comment by alexrockhill** [comment]
The other convenience function I would like to suggest is `autoreject.transform(epochs, interp_thresh=0.5)` this could be `None` by default in which case nothing would change (current behavior). If a float is passed, then if more than that much of the channel is marked as bad or for interpolation, the whole channel is interpolated (since presumably it should have been marked as bad and drop before autoreject if it wasn't supposed to be interpolated).

I think this would be a handy convenience and good practice for internal consistency (as above) but again, totally defer to your opinions on whether this is too much post-hoc meddling.

**Comment by alexrockhill** [comment]
@jasmainak, what do you think re the second suggestion? If you'd rather not open that much post-hoc modification that's fine, I totally defer to your preference

**Comment by jasmainak** [comment]
I think it needs a better name to convey the aspect of epochs/time :) But otherwise it's okay. 

Now note that the idea of autoreject is to work "out of the box" ... so by default users shouldn't have to tweak parameters, as otherwise it opens a pandora's box. You'll have to tweak these parameters for different subjects/runs etc. You'd need to possibly do some science to figure out what values of these parameters are ideal across datasets.
